src/mapreduce.jl

```diff
@@ -0,0 +1,276 @@
+function tmapreduce end
+
+@doc _make_docstring(:reduce, ["op", "src::AbstractArray"],  ["[init]"]) ->
+function treduce end
+
+############################## Helper functions ##############################
+struct Batches{V}
+    firstindex::Int
+    lastindex::Int
+    batch_size::Int
+    values::V
+    length::Int
+end
+
+function Batches(values, batch_size::Integer)
+    @argcheck batch_size > 0
+    r = eachindex(values)
+    @assert r isa AbstractUnitRange
+    len = ceil(Int, length(r) / batch_size)
+    Batches(first(r), last(r), batch_size, values, len)
+end
+
+Base.length(o::Batches) = o.length
+Base.eachindex(o::Batches) = Base.OneTo(length(o))
+function Base.getindex(o::Batches, i)
+    @boundscheck (@argcheck i in eachindex(o))
+    start = o.firstindex + (i-1) * o.batch_size
+    stop  = min(start + (o.batch_size) -1, o.lastindex)
+    o.values[start:stop]
+end
+
+function default_batch_size(len)
+    len <= 1 && return 1
+    nthreads=Threads.nthreads()
+    items_per_thread = len/nthreads
+    items_per_batch = items_per_thread/4
+    clamp(1, round(Int, items_per_batch), len)
+end
+
+function Base.iterate(o::Batches, state=1)
+    if state in eachindex(o)
+        o[state], state+1
+    else
+        nothing
+    end
+end
+
+mutable struct _RollingCutOut{A,I<:AbstractUnitRange,T} <: AbstractVector{T}
+    array::A
+    eachindex::I
+end
+
+function _RollingCutOut(array::AbstractArray, indices)
+    T = eltype(array)
+    A = typeof(array)
+    I = typeof(indices)
+    _RollingCutOut{A, I, T}(array, indices)
+end
+
+Base.size(r::_RollingCutOut) = (length(r.eachindex),)
+
+function Base.eachindex(r::_RollingCutOut, rs::_RollingCutOut...)
+    for r2 in rs
+        @assert r.eachindex == r2.eachindex
+    end
+    Base.IdentityUnitRange(r.eachindex)
+end
+Base.axes(r::_RollingCutOut) = (eachindex(r),)
+
+@inline function Base.getindex(o::_RollingCutOut, i)
+    @boundscheck checkbounds(o, i)
+    @inbounds o.array[i]
+end
+
+@inline function Base.setindex!(o::_RollingCutOut, val, i)
+    @boundscheck checkbounds(o, i)
+    @inbounds o.array[i] = val
+end
+
+############################## tmap, tmap! ##############################
+struct MapWorkspace{F,B,AD,AS}
+    f::F
+    batches::B
+    arena_dst_view::AD
+    arena_src_views::AS
+end
+
+@noinline function run!(o::MapWorkspace)
+    let o=o
+        Threads.@threads for i in 1:length(o.batches)
+            tid = Threads.threadid()
+            dst_view  = o.arena_dst_view[tid]
+            src_views = o.arena_src_views[tid]
+            inds = o.batches[i]
+            dst_view.eachindex = inds
+            for view in src_views
+                view.eachindex = inds
+            end
+            Base.map!(o.f, dst_view, src_views...)
+        end
+    end
+end
+
+function create_arena_src_views(srcs, sample_inds)
+    nt = Threads.nthreads()
+    [Base.map(src -> _RollingCutOut(src, sample_inds), srcs) for _ in 1:nt]
```

Comment:
Can we use a global constant for the number of threads and make this a tuple to avoid allocating?

Reply:
Is `Threads.nthreads()` is a constant? Is this documented somewhere?
Also note that while the compiler is very good at eliding allocation of tuples of bittypes, tuples of non bittypes typically allocate. (and srcs are very likely non bittypes.)

Reply:
`Threads.nthreads()` is set with an environment variable before opening Julia, at least until Julia 1.1. Not quite sure about the newer versions of Julia in the making. So yes it is a constant.

Reply:
Ok then why is `Threads.nthreads()` a function and not a constant in the first place? ALso how would making this a tuple interact with precompilation if the env variable gets changed between julia sessions?